### PR TITLE
Adds a version numbering proposal for the WPILib Suite.

### PR DESCRIPTION
**Pull request by 333fred**
This proposal is a major change for how we do versioning, to make it much more sane both for us and teams. I've had these ideas swirling around in my head for several months, but have finally gotten them written down in what I think is a coherent and ordered form. I welcome input from anyone on the team, but I especially want to call out @bradamiller, @Kevin-OConnor, and @PeterJohnson to take a look at this as the Control System heads. Note that at the very least, some work needs to be done to get releases up and running again, as we currently don't have the release system set up. I'd like to get this reviewed this week so I can begin implementation next week, targeted to finish by the end of August.

### CODE REVIEW

**Comment by PeterMitrano** [comment]
everything without any I don't have a very informed opinion, but LGTM. My only question is, do any of the files in C++/Java include/import the auto-generated version file? or is it simply packaged along with the jar/library? If no one includes it, then we could regenerate it every time without any of the actual source needing to be recompiled

**Comment by 333fred** [comment]
From @ThadHouse on Slack: 
I think the versioning should be semver based
[13:58]  
I know it results in basically an unused number in our case, but I think thats ok.
Actually we could use the SemVer to state exceptional cases where breaking changes had to be made mid season
[14:03]  
Like `release-year.breaking_changes.release_number-suffix`. I would say even on a breaking change increase the release_number, that way the only thing teams really need to see is that 3rd number and the first number. the second number can be for only teams that care about things breaking

**Comment by 333fred** [comment]
From @PeterJohnson on Slack:
Arguably, we could make the HAL separately versioned from WPILib.  The HAL will need to track NI versions but WPILib might not.  Not sure if that's worth it though or whether it's more of a YAGNI thing

**Comment by 333fred** [comment]
@ThadHouse, I considered this, but I'm not a fan. I think that, for the majority of teams, they don't care about that other version number, just the year of the release, and the release number. I'm also not a fan of the idea that we kind of use SemVer, but not really. That would effectively leave gaps in our version history, which is relatively unexpected. Note that I'm only proposing this system for the parts of the library that are intrinsically tied to FRC. Things that aren't versioned on a year to year system like ntcore are SemVer.

**Comment by 333fred** [comment]
@PeterJohnson, yeah, I think that's YAGNI. The HAL isn't even versioned to begin with right now, as it doesn't have internal versions and isn't published as a maven artifact.

**Comment by 333fred** [comment]
From @PeterJohnson on Slack:
Re @PeterMitrano comment, if the version number was in a cpp file and just extern'd from the header, that would avoid recompiling other files when it changes...

**Comment by 333fred** [comment]
@PeterMitrano, everywhere that uses it now will import/include those files, namely RobotBase. That being said, to republish the eclipse plugins, we'd have to unpack the plugin, unpack the C++/Java resources zips, unpack the .a/.jar files, patch them, recalculate checksums, repack the resource zips, repack the eclipse plugins, recalculate checksums both inside the jar and in the eclipse update site itself, and I'm probably missing steps.

**Comment by ThadHouse** [comment]
What if local builds always just used a checked in version file, and jenkins was the only think that updated the version file, but it wouldn't commit it.

**Comment by 333fred** [comment]
I'm not sure what you mean by that. If nothing's committed, it can't know what version to publish.

**Comment by ThadHouse** [comment]
Oh nevermind I was reading the drawbacks wrong.

**Comment by PeterJohnson** [comment]
One thing we should do at the same time is have the builds pull ithe version from the git version number (eg tag), ala git describe.

**Comment by virtuald** [comment]
Seems sane to me.

**Comment by Kevin-OConnor** [comment]
LGTM

**Comment by 333fred** [comment]
@PeterJohnson, what process do you view this as being? One thing that I'm hoping to do with this system is make everything obvious, and I can't come up with a scenario involving git tags that makes everything obvious and simple.

**Comment by 333fred** [comment]
And generated for consistency.

**Comment by PeterJohnson** [comment]
Instead of reading from a local JSON file, you run something like `git describe --match "v[0-9]*" --abbrev=4 HEAD` to get the version information (you can still have the JSON file option as an override if desired).  This allows you to do version number bumps by adding a specially formatted git tag (e.g. `v2016.1.0`) rather than editing/committing a file change.  One nice thing is that git describe does automatic sub-tier versioning (the number of commits since the last tag gets tacked on to become e.g. `2016.1.0.25`).  I don't have a gradle example at hand, but examples of how to automate this:
- Shell: https://github.com/yasm/yasm/blob/master/YASM-VERSION-GEN.sh
- Batch file: https://github.com/yasm/yasm/blob/master/YASM-VERSION-GEN.bat
- CMake: https://github.com/yasm/yasm-nextgen/blob/master/cmake/modules/VersionGen.cmake

**Comment by PeterJohnson** [comment]
Gradle examples...
One way: http://mrhaki.blogspot.com/2015/04/gradle-goodness-use-git-commit-id-in.html
A plugin: https://github.com/palantir/gradle-git-version
Various other approaches: http://stackoverflow.com/questions/17097263/automatically-versioning-android-project-from-git-describe-with-android-studio-g
Other plugins: https://plugins.gradle.org/search?term=git+describe

**Comment by 333fred** [comment]
I've had a couple of thoughts around the FIRST versioning scheme, and I'd like some input before I modify the doc, especially from @Kevin-OConnor. I've noticed a potential problem with teams compiling their own version of WPILib during the season:
1. Our release at the time is `2017.1`. Therefore, the team's version during compile will be `2017.2-alpha-buildstamp`.
2. We release a new mandatory version. This increments our number to `2017.2`.
3. `2017.2` and `2017.2-alpha` don't look that different. I can see an inspector getting confused, especially if we allow `-alpha` for teams who compile their own WPILib. The team is allowed to compete, even though they don't have the required version.

I have 2 ideas to address this, not mutually exclusive. First, we start by hewing closer to SemVer. It's still not official SemVer, however, as the actual semantics are still different. Something like `competition-year.required-version.optional-version`. So the first release of the year is `2017.1.1`. We release an optional update during week 1, which is `2017.1.2`. We then discover a major bug, and release a new required version. This is `2017.2.1`. I think this is far easier for inspectors to look at (any version of 2017.2 is easier than the aforementioned example), and easier for teams as well.
The second piece is creating a label for team-built libraries. This would take the same form as the alpha builds, except instead of `-alpha-buildsnapshot`, it would be `-team-buildsnapshot`. Kevin, I'm not sure what we would want to do for legal versioning (would alpha be legal or not?), but at the very least, this won't increment the build number like an alpha version would. For example, lets say that the version is currently `2017.1.1`. New builds would be `2017.1.2-alpha-snapshot` by default, unless built with a `-PteamBuild` (or something to that effect) flag, in which case it would be `2017.1.1-team-snapshot`. Thoughts?

**Comment by PeterJohnson** [comment]
This is why using git describe is good.  If you have local commits against the git branch, you end up with a version number like `2017.2.1.<#of commits>` (e.g. `2017.2.1.2`), if you have uncommitted changes it's `2017.2.1.<# of commits>-dirty` (e.g. `2017.2.1.2-dirty` or `2017.2.1-dirty` if there's no commits).  This of course depends on people actually using git... but people who really are building from source can override all of this anyway.

**Comment by 333fred** [comment]
@PeterJohnson, I think we're both misunderstanding each other. In what scenario would a git describe world result in `2017.2.1.x`? That could only happen if the latest git tag is `2017.2.0`, unless I'm not getting something you're saying.

**Comment by PeterJohnson** [comment]
`git-describe --dirty` results in `<last tag>-<# of commits since last tag>-<latest commit sha hash>`, optionally followed by `-dirty`, e.g. `v1.3.0-28-g51af408-dirty` if the last tag is `v1.3.0`, 28 commits have been made since then, the last commit sha ends with `g51af408`, and the tree is dirty (uncommitted changes).  See https://git-scm.com/docs/git-describe.  We could of course change the formatting of this with scripting.

**Comment by PeterJohnson** [comment]
So in our case, if we tag with `v2017.2.0`, `v2017.2.1`, etc, any user changes would be `v2017.2.0-xx` if they haven't merged 2017.2.1, `v2017.2.1-xx` if they have, etc.

**Comment by 333fred** [comment]
Again, I think we're both missing something here. A tag of `2017.1.0` should result in a local build of `2017.1.1-alpha-snapshot`, not `2017.1.0-alpha-snapshot`. Why would we continue development of `2017.1.0` after it's been tagged?

**Comment by PeterJohnson** [comment]
Why?  if you've made no changes, you should get the same version number.

**Comment by 333fred** [comment]
But the scenario here is if you have made changes. I find it weird that I make some commits and it wouldn't be work towards the next version.

**Comment by PeterJohnson** [comment]
It gets added on as a sub minor change.  So release is 2017.1.0, once you've made changes it becomes 2017.1.0-1 (or 2017.1.0.1), -2 or .2, ... Once you tag a new release of 2017.1.1 the sequence restarts.

**Comment by 333fred** [comment]
How do alpha/beta/rc fit in here? Does git-describe understand that?

**Comment by PeterJohnson** [comment]
If you tag those it will work fine.  Any tag starting with v I think works

**Comment by 333fred** [comment]
Alright, I'm mostly convinced here. One change from your proposal though: in order to make sure Maven keeps working as expected, the number of commits can't be `2017.1.0.x`. Otherwise, if I working on a branch that's 2 commits ahead, and then I switch to a branch with 1 commit ahead and try to build a downstream dependency, it'll pick up the one later by date, and not later by commit number (as 2017.1.0.2-dirty would be considered a more recent build than 2017.1.0.1-dirty).

**Comment by 333fred** [comment]
We'll put the number of commits after the date, so something like `2017.1.0-timestamp-x-dirty`

**Comment by 333fred** [comment]
I'm also still of the opinion that having a team flag for team builds would be convenient for those few who recompile.

**Comment by PeterJohnson** [comment]
Isn't the underlying issue that there's no easy way to clean up locally published artifacts?  If you could, then the issue you describe wouldn't be a problem.  However, in the absence of a fix for that issue, I'm okay with the timestamp-based approach.

**Comment by 333fred** [comment]
Yeah, there is no easy way to clean them up. Cleaning up the repo should be a user-initiated action, as they might want to keep builds around for some reason or another.

**Comment by PeterJohnson** [comment]
One concern I have with any kind of timestamp based approach, however, is how it triggers rebuilds.  Does this mean every "gradlew build" will force a recompile of version.cpp (or worse, a widely included version.h), relink, rezip, etc, even if no files have otherwise changed?  Or does that only happen when publish is run?  I think the latter would be far preferable if we could get it to work.

**Comment by PeterJohnson** [comment]
Or maybe the timestamp-based version is only part of the package name, not part of the internal version number?  There's potential downsides to that too, of course.

**Comment by 333fred** [comment]
Hmm, that's a wrinkle I hadn't considered that much. I'll need to think about how we can address that for a minute.

**Comment by 333fred** [comment]
In some ways this is a chicken and egg problem. We don't want to generate a new version number if there's been no changes, but how can we determine if there's no changes without building twice...

**Comment by PeterJohnson** [comment]
If we only regen the version number on publish, will that force a rebuild only on publish?  Or is publish special in such a way that won't work?

**Comment by 333fred** [comment]
Well, that would require foreknowledge that publish will occur. I'm looking into whether this is possible to detect now.

**Comment by PeterJohnson** [comment]
I suppose we could create a "prepublish" task that does it, but it would be much better if it ran automatically as part of running publish!  Or could we make a new task that depends on both prepublish and publish and thus did the right thing with a single command (which unfortunately is not called publish)

**Comment by 333fred** [comment]
It looks like we can watch the task graph, and see if `publish` is added, and add that task if it's there.

**Comment by 333fred** [comment]
Alright, so here's the logic I'm thinking for this:
- If no number has been generated at all (generated files are missing), a build number is generated regardless of publish status.
- If a number does exist, and publish will not be executed, the number will not be regenerated.
- If a number does exist, and publish will be executed, the number will be regenerated.

**Comment by PeterJohnson** [comment]
Makes sense to me.

**Comment by 333fred** [comment]
@PeterJohnson, @Kevin-OConnor, @bradamiller, I've addressed the PR feedback. I think this should be ready to go, please take a look.

File: Version Generation.adoc [hunk 1]
```diff
@@ -24,19 +24,31 @@ This design covers 2 main changes to WPILib's build system: version number code
...
 [source,c++]
 ----
 /* BSD License Header */
 #pragma once
 
 /*
- * Autogenerated file! Do not manually edit this file. To change the version number,
- * change the WPILib version file located at <allwpilib_root>/wpilib_version.json
+ * Autogenerated file! Do not manually edit this file. This version is regenerated
+ * any time the publish task is run, or when this file is deleted.
```
Comment (PeterJohnson) [inline Version Generation.adoc]: Why is the header file regenerated?  It should never change...

Comment (333fred) [reply Version Generation.adoc]: While true, I guess I just find the idea of having half of this checked in while the other half is generated weird.

Comment (PeterJohnson) [reply Version Generation.adoc]: It's pretty common to have a header refer to functions in a generated file, for example.  And unless you're careful to not actually write the file if it hasn't changed (which it never will), you will cause spurious rebuilds of other files that include version.h.  Seems easier to just check it in and not worry about it.

**Comment by 333fred** [comment]
@PeterJohnson, are you all set with this? If so, I'll merge it in.

**Comment by PeterJohnson** [comment]
LGTM

**Pull request merged**